src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,191 @@
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class StringBuilderAppendShouldNotTakeSubstringFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray
+            .Create(
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId,
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdTwoParameterId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                if (diagnostic.Id == StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId)
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadFix,
+                                ct => FixCodeOneParameter(context.Document, context.Span, ct)),
+                            diagnostic);
+                }
+                else
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadFix,
+                                ct => FixCodeTwoParameters(context.Document, context.Span, ct)),
+                            diagnostic);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static async Task<Document> FixCodeOneParameter(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var nodeToFix = root.FindNode(span, getInnermostNodeForTie: true);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var typedNodeToFix =  semanticModel.GetOperation(nodeToFix) as IInvocationOperation;
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var fixComponents = GetFixComponents(generator, typedNodeToFix);
+
+            // generate text.Length (or whatever "text" is instead in the actual code)
+            var lengthNode = generator
+                .MemberAccessExpression(
+                    fixComponents.StringArgument.Syntax, 
+                    generator.IdentifierName(nameof(string.Length)));
```

Comment:
do we have a test where the string argument is complex (i.e. not an identifier/literal)?  it will help ensure that this is parenthesized properly.